File: frontend/zidify_app/lib/features/user/domain_layer/entities/user_entity.dart
class UserEntity {
  String id;
  String email;
  String firstName;
  String lastName;
  String phoneNumber;
  dynamic dob;
  dynamic address;
  dynamic identificationType;
  dynamic identificationNumber;
  dynamic nationality;
  dynamic occupation;
  dynamic googleId;
  dynamic avatar;
  bool isEmailVerified;
  bool isPhoneNumberVerified;
  bool verified;
  String refreshToken;
  dynamic emailVerificationToken;
  dynamic emailVerificationTokenExpiry;
  dynamic passwordChangedAt;
  dynamic passwordResetOtp;
  dynamic passwordResetOtpExpiry;
  DateTime createdAt;
  DateTime updatedAt;
  UserSavingsTotalEntity userSavingsTotal;
  List<ActivityEntity> activities;

  UserEntity({
    required this.id,
    required this.email,
    required this.firstName,
    required this.lastName,
    required this.phoneNumber,
    required this.dob,
    required this.address,
    required this.identificationType,
    required this.identificationNumber,
    required this.nationality,
    required this.occupation,
    required this.googleId,
    required this.avatar,
    required this.isEmailVerified,
    required this.isPhoneNumberVerified,
    required this.verified,
    required this.refreshToken,
    required this.emailVerificationToken,
    required this.emailVerificationTokenExpiry,
    required this.passwordChangedAt,
    required this.passwordResetOtp,
    required this.passwordResetOtpExpiry,
    required this.createdAt,
    required this.updatedAt,
    required this.userSavingsTotal,
    required this.activities,
  });
}

class UserSavingsTotalEntity {
  String id;
  String userId;
  int totalSaveGoals;
  int totalLockBoxes;
  int totalSaveBox;
  int overallTotal;
  DateTime createdAt;
  DateTime updatedAt;

  UserSavingsTotalEntity({
    required this.id,
    required this.userId,
    required this.totalSaveGoals,
    required this.totalLockBoxes,
    required this.totalSaveBox,
    required this.overallTotal,
    required this.createdAt,
    required this.updatedAt,
  });
}

class ActivityEntity {
  String id;
  String userId;
  int amount;
  String category;
  String description;
  DateTime createdAt;
  DateTime updatedAt;

  ActivityEntity({
    required this.id,
    required this.userId,
    required this.amount,
    required this.category,
    required this.description,
    required this.createdAt,
    required this.updatedAt,
  });
}

File: backend/src/users/entities/user.entity.ts
import {
  ROLE,
  SaveBox,
  LockBox,
  SaveGoal,
  Activity,
  FundDestination,
  FundingSource,
  UserSavingsTotal,
} from '@prisma/client';

export class User {
  id: string;
  email: string;
  firstName: string;
  lastName: string;
  password?: string;
  phoneNumber?: string;

  dob?: Date;
  address?: string;
  identificationType?: string;
  identificationNumber?: string;
  nationality?: string;
  occupation?: string;

  googleId?: string;
  avatar?: string;
  isEmailVerified: boolean = false;
  isPhoneNumberVerified: boolean = false;

  verified: boolean = false;
  refreshToken?: string;
  emailVerificationToken?: string;
  emailVerificationTokenExpiry?: Date;
  passwordChangedAt?: Date;
  passwordResetOtp?: number;
  passwordResetOtpExpiry?: Date;

  role: ROLE = ROLE.user;

  saveBox?: SaveBox;
  lockBoxes: LockBox[] = [];
  saveGoals: SaveGoal[] = [];

  paymentMethods: FundingSource[] = []; // FundingSource[]
  userSavingsTotal?: UserSavingsTotal; // UserSavingsTotal
  activities: Activity[] = []; // Activity[]
  fundDestinations: FundDestination[] = []; // FundDestination[]

  createdAt: Date = new Date();
  updatedAt: Date = new Date();

  constructor(data: Partial<User> = {}) {
    Object.assign(this, data);
  }
}
